### DISCUSSION TITLE
Sense amplifiers for reading FGFETs

### DISCUSSION THREAD

**Discussion by MikeHaydell** [hneemann/Digital#1130] [Q&A] [answered]
I'm trying to make a custom non-volatile memory element and have sorted out the erase/write operations, but am having trouble with reading the high z state of a programmed floating gate FET as logical 0. I know real world devices use sense amplifiers and comparators to measure the current, but every design I try to implement gives me errors. Is there a workaround for this (other than using a pull down resistor) or are analog circuits not possible (I mean it is Digital after all:)

**Comment by hneemann** [comment, marked as answer]
It is possible to sense high-z:

![highz](https://user-images.githubusercontent.com/7190563/224631392-5347e63b-8c1c-4858-9da0-320ac3ab2446.png)

In reality this would not work, because the circuit relies on the fact that a transistor, which has a floating input, does not conduct. In reality, a floating input would drift, and possibly one of the transistors or even both would become conductive.

**Comment by MikeHaydell** [comment]
![FGFET](https://user-images.githubusercontent.com/50136742/224633661-39bd3aaf-8149-464e-9355-6eec18842cb8.png)

Wow you're fast at replying. Thanks!

Maybe I should have included the above image for clarity. But yeah, your reply is exactly my issue. I want to read the FGFET as 1 when its not programmed and 0 when it is programmed. The high-z state gives floating inputs (like in image A) for Bus0 and Bus2. The pull down resistors (image B) solve the problem (in Digital) but I understand a real circuit would have issues with that which is why they typically employ sense amplifiers. I'm trying to be as realistic as possible (which is why I switched from logisim to Digital) but when I try to use a sense amplifier (based on the differential sense amplifier wikipedia page with current mirror, below image) I get oscillations. I'm guessing this is either the improper use/design of a sense amplifier or Digital has issues with this circuit.

![Picture2](https://user-images.githubusercontent.com/50136742/224644799-236ad84b-944c-4909-a1c5-4b51457a2da4.png)

**Reply by hneemann** [reply]
This circuit contains a current mirror with a differential amplifier. Something like this can only be simulated with an analog simulator.

**Reply by MikeHaydell** [reply]
Ok well at least I know now. Thanks!
